### PR TITLE
Add support for imagePullSecrets

### PR DESCRIPTION
**Pull request by garonsky**
Add support for imagePullSecrets in the CDI CR to support pulling images from repositories that require secrets.

The imagePullSecrets is propagated to the following components: cdi-apiserver, cdi-deployment, and cdi-uploadproxy. The definition of imagePullSecrets in cdi-operator must be done manually.

**Does this PR introduce a user-facing change?**

```release-note
Added support for imagePullSecrets to the CDI custom resource and the cdi-operator resource.  Updates to the imagePullSecrets value in the custom resource are automatically propagated to cdi-apiserver, cdi-deployment, and cdi-uploadproxy.  The cdi-operator resource should be manually updated.
```

### CODE REVIEW

**Comment by mhenriks** [comment]
thanks for the pr @garonsky!

/ok-to-test

**Comment by mhenriks** [comment]
@garonsky have you seen the discussion in https://github.com/kubevirt/containerized-data-importer/issues/2395 on support for secrets in "worker" pods?  Mind sharing your thoughts there?  Looks like you intend to support clone source pods in this pr?  What are your thoughts on the other workers?

File: staging/src/kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1/types.go [hunk 1]
```diff
@@ -653,7 +653,9 @@ type CDICertConfig struct {
 type CDISpec struct {
 	// +kubebuilder:validation:Enum=Always;IfNotPresent;Never
 	// PullPolicy describes a policy for if/when to pull a container image
-	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty" valid:"required"`
+	ImagePullPolicy  corev1.PullPolicy             `json:"imagePullPolicy,omitempty" valid:"required"`
+	// The imagePullSecrets to pull the container images
+	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
```
Comment (mhenriks) [inline staging/src/kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1/types.go:658]: I assume that we want regular users to know the names of secrets we expect them to create?  So I suggest moving `ImagePullSecrets` to `CDIConfig`.  So CDI will look like this:

```yaml
...
kind: CDI
spec:
  config:
    imagePullSecrets:
    - secret-name
...
```

And CDIConfig should have `imagePullSecrets` in the status like so:

```yaml
...
kind: CDIConfig
metadata:
  name: config
spec:
  imagePullSecrets:
    - secret-name
status:
  imagePullSecrets:
    - secret-name
...
```

CDIConfig has a singleton (named config) that is created by the config-controller and readable by all users.  The `spec` section is `spec.config` from the CDI resource.

I refer you again to how `PodResourceRequirements` are handled

**Review by mhenriks — COMMENTED** [review]
@garonsky thanks for the PR, great start!  Looks like we mostly just have to flush out the support for worker pods and.

**Comment by garonsky** [comment]
@mhenriks thank you for the great feedback!  I tried to implement the suggestions, let me know if this is more of what you had in mind.

**Comment by mhenriks** [comment]
@garonsky I did a quick scan and I think this looks great!  I will dig a little deeper tomorrow and keep you posted.  Sorry for the delayed response.  Been on PTO.

File: pkg/controller/import-controller.go [hunk 2]
```diff
@@ -867,6 +868,8 @@ func createImporterPod(log logr.Logger, client client.Client, args *importerPodA
 		return nil, err
 	}
 
+	args.imagePullSecrets = cdiv1.CDIConfig{}.Spec.ImagePullSecrets
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:871]: Should be calling `GetImagePullSecrets` here, right?

**Review by mhenriks — COMMENTED** [review]
Great work!  Couple places that need pull secrets:

https://github.com/garonsky/containerized-data-importer/blob/add_imagePullSecrets/pkg/controller/dataimportcron-controller.go#L877-L883

https://github.com/garonsky/containerized-data-importer/blob/add_imagePullSecrets/pkg/controller/datavolume/pvc-clone-controller.go#L1246-L1258

https://github.com/garonsky/containerized-data-importer/blob/add_imagePullSecrets/pkg/controller/import-controller.go#L1082-L1092

I think that's it!

**Comment by garonsky** [comment]
@mhenriks Thank you so much for the guidance!  I made some changes.  Let me know if I captured your suggestions with the latest changes.

File: pkg/controller/config-controller.go [hunk 3]
```diff
@@ -232,6 +236,14 @@ func (r *CDIConfigReconciler) reconcileStorageClass(config *cdiv1.CDIConfig) err
 	return nil
 }
 
+func (r *CDIConfigReconciler) reconcileImagePullSecrets(config *cdiv1.CDIConfig) error {
+	if config.Spec.ImagePullSecrets != nil {
+		config.Status.ImagePullSecrets = config.Spec.ImagePullSecrets
+	}
```
Comment (mhenriks) [inline pkg/controller/config-controller.go:242]: `config.Status.ImagePullSecrets` will not get unset if `config.Spec.ImagePullSecrets` deleted

**Review by mhenriks — COMMENTED** [review]
@garonsky seems to work well, just one bug I found

**Comment by mhenriks** [comment]
/lgtm
/approve

Thanks @garonsky!

**Comment by aglitke** [comment]
Fixes #2395

**Pull request merged**